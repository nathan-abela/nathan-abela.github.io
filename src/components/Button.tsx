import styled, { css, keyframes } from 'styled-components';
import {
    color,
    compose,
    layout,
    padding,
    variant,
    ColorProps,
    LayoutProps,
    PaddingProps
} from 'styled-system';

interface ButtonProps {
    variant?: 'primary' | 'secondary' | 'animated';
}

const animation = keyframes`
    0% {
        background-position: 0 50%;
    }

    20% {
        background-position: 35% 80%;
    }

    40% {
        background-position: 100% 50%;
    }

    60% {
        background-position: 70% 0;
    }

    80% {
        background-position: 70% 0;
    }

    100% {
        background-position: 0 50%;
    }
`;

const Button = styled.button<
    ButtonProps & ColorProps & LayoutProps & PaddingProps
    >`
    font-size: 1rem;
    font-weight: bold;
    line-height: 1.5rem;

    display: inline-block;
    align-items: center;
    justify-content: center;

    margin: 0;
    padding: 20px 50px;

    cursor: pointer;
    transition: ease-in-out 0.15s;

    color: white;
    border: none;
    border-radius: 12px;
    outline: 0;

    ${({ variant }) =>
        variant === 'animated' &&
        css`
            animation: ${animation} 2s linear infinite;
        `}
    ${compose(padding, color, layout)}

    ${variant({
            variants: {
                animated: {
                    background:
                        'linear-gradient(200deg, rgb(255, 95, 109),rgb(255, 195, 113))',
                },
                primary: {
                    color: 'white',
                    backgroundColor: '#050505',
                    '&:hover': {
                        color: 'white',
                        backgroundColor: 'black',
                        transform: 'scale(1.05)',
                    },
                },
                secondary: {
                    color: 'black',
                    backgroundColor: '#f5f5f5',
                    '&:hover': {
                        color: 'white',
                        backgroundColor: 'black',
                        transform: 'scale(1.05)',
                    },
                },
            },
        })};
`;

Button.defaultProps = {
    variant: 'primary',
};

export default Button;
