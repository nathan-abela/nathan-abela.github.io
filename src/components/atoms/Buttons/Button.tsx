import { FC, ReactNode } from 'react';
import { StyledButton, StyledLink } from './styles';

export interface ButtonProps {
	href?: string;
	onClick?: (...args: any) => void;
	target?: '_blank' | '_self' | '_parent';
	children?: ReactNode;
	transparent?: boolean;
}

const Button: FC<ButtonProps> = ({ children, href, target, onClick, transparent }) => {
	if (href) {
		return (
			<StyledLink href={href} target={target} transparent={transparent}>
				{children}
			</StyledLink>
		);
	}

	return <StyledButton onClick={onClick}>{children}</StyledButton>;
};

export { Button };
