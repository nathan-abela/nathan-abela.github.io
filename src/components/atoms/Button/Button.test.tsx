import { render } from '@testing-library/react';

import { Button } from '../Button';

describe('Button', () => {
	it('renders with link correctly', () => {
		const { container } = render(
			<Button href="/url" target="_blank">
				Button text
			</Button>
		);
		expect(container).toMatchSnapshot();
	});

	it('renders correctly', () => {
		const { container } = render(
			// eslint-disable-next-line prettier/prettier
			<Button onClick={jest.fn}>Button text</Button>
		);
		expect(container).toMatchSnapshot();
	});
});
