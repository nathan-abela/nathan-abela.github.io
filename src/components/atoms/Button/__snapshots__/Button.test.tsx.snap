// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<div>
  <button
    class="styles__StyledButton-sc-1tgmejn-0 iFnVFy"
  >
    Button text
  </button>
</div>
`;

exports[`Button renders with link correctly 1`] = `
<div>
  <a
    class="styles__StyledLink-sc-1tgmejn-1 dogzke"
    href="/url"
    target="_blank"
  >
    Button text
  </a>
</div>
`;
